### ISSUE TITLE
MissingMethodException: TypeBiulder.DefineProperty, on Unity3D/Mono

### ISSUE THREAD

**Issue by dtchepak** [castleproject/Core#319]
Hi Castle Core!

Not sure you'll be able to help with this but thought I'd raise it in case you can give me a lead on how to chase this up. An NSubstitute user [has reported a problem](https://twitter.com/GameDevAlex/status/913678227176148992) trying to proxy a type with properties on it, getting the following exception:

```
System.MissingMethodException : Method not found: 'System.Reflection.Emit.TypeBuilder.DefineProperty'.
---
at Castle.DynamicProxy.Generators.Emitters.AbstractTypeEmitter.CreateProperty (System.String name, PropertyAttributes attributes, System.Type propertyType, System.Type[] arguments) [0x00000] in <filename unknown>:0 
  at Castle.DynamicProxy.Generators.MetaProperty.BuildPropertyEmitter (Castle.DynamicProxy.Generators.Emitters.ClassEmitter classEmitter) [0x00000] in <filename unknown>:0 
  at Castle.DynamicProxy.Contributors.CompositeTypeContributor.ImplementProperty (Castle.DynamicProxy.Generators.Emitters.ClassEmitter emitter, Castle.DynamicProxy.Generators.MetaProperty property, Castle.DynamicProxy.ProxyGenerationOptions options) [0x00000] in <filename unknown>:0 
  at Castle.DynamicProxy.Contributors.CompositeTypeContributor.Generate (Castle.DynamicProxy.Generators.Emitters.ClassEmitter class, Castle.DynamicProxy.ProxyGenerationOptions options) [0x00000] in <filename unknown>:0 
  at Castle.DynamicProxy.Generators.InterfaceProxyWithoutTargetGenerator.GenerateType (System.String typeName, System.Type proxyTargetType, System.Type[] interfaces, INamingScope namingScope) [0x00000] in <filename unknown>:0 
  at Castle.DynamicProxy.Generators.InterfaceProxyWithTargetGenerator+<>c__DisplayClass1.<GenerateCode>b__0 (System.String n, INamingScope s) [0x00000] in <filename unknown>:0 
  at Castle.DynamicProxy.Generators.BaseProxyGenerator.ObtainProxyType (Castle.DynamicProxy.Generators.CacheKey cacheKey, System.Func`3 factory) [0x00000] in <filename unknown>:0 
  at Castle.DynamicProxy.Generators.InterfaceProxyWithTargetGenerator.GenerateCode (System.Type proxyTargetType, System.Type[] interfaces, Castle.DynamicProxy.ProxyGenerationOptions options) [0x00000] in <filename unknown>:0 
  at Castle.DynamicProxy.DefaultProxyBuilder.CreateInterfaceProxyTypeWithoutTarget (System.Type interfaceToProxy, System.Type[] additionalInterfacesToProxy, Castle.DynamicProxy.ProxyGenerationOptions options) [0x00000] in <filename unknown>:0 
  at Castle.DynamicProxy.ProxyGenerator.CreateInterfaceProxyTypeWithoutTarget (System.Type interfaceToProxy, System.Type[] additionalInterfacesToProxy, Castle.DynamicProxy.ProxyGenerationOptions options) [0x00000] in <filename unknown>:0 
  at Castle.DynamicProxy.ProxyGenerator.CreateInterfaceProxyWithoutTarget (System.Type interfaceToProxy, System.Type[] additionalInterfacesToProxy, Castle.DynamicProxy.ProxyGenerationOptions options, Castle.DynamicProxy.IInterceptor[] interceptors) [0x00000] in <filename unknown>:0 
  at NSubstitute.Proxies.CastleDynamicProxy.CastleDynamicProxyFactory.CreateProxyUsingCastleProxyGenerator (System.Type typeToProxy, System.Type[] additionalInterfaces, System.Object[] constructorArguments, IInterceptor interceptor, Castle.DynamicProxy.ProxyGenerationOptions proxyGenerationOptions) [0x00000] in <filename unknown>:0 
  at NSubstitute.Proxies.CastleDynamicProxy.CastleDynamicProxyFactory.GenerateProxy (ICallRouter callRouter, System.Type typeToProxy, System.Type[] additionalInterfaces, System.Object[] constructorArguments) [0x00000] in <filename unknown>:0 
  at NSubstitute.Proxies.ProxyFactory.GenerateProxy (ICallRouter callRouter, System.Type typeToProxy, System.Type[] additionalInterfaces, System.Object[] constructorArguments) [0x00000] in <filename unknown>:0 
  at NSubstitute.Core.SubstituteFactory.Create (System.Type[] typesToProxy, System.Object[] constructorArguments, SubstituteConfig config) [0x00000] in <filename unknown>:0 
  at NSubstitute.Core.SubstituteFactory.Create (System.Type[] typesToProxy, System.Object[] constructorArguments) [0x00000] in <filename unknown>:0 
  at NSubstitute.Substitute.For (System.Type[] typesToProxy, System.Object[] constructorArguments) [0x00000] in <filename unknown>:0 
  at NSubstitute.Substitute.For[IInControlInputService] (System.Object[] constructorArguments) [0x00000] in <filename unknown>:0 
  at Tests.Test_InputManager.setup () [0x00008] in /Users/dtchepak/Downloads/DarkGame/Assets/Editor/Tests/Player/Test_InputManager.cs:25 
  at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x000d0] in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection/MonoMethod.cs:222
```

The proxy creation is done via `InputService = Substitute.For<IInControlInputService>();`, and the type being proxied looks like this:

```
public interface IInControlInputService
{
	Vector2 LeftStick { get; }
	Vector2 RightStick { get; }
	bool GetControlIsDown(InputControlType action);
}
```

If I remove the `LeftStick` and `RightStick` properties (or replace them with methods) the proxy is created fine, so it seems to be that only `System.Reflection.Emit.TypeBuilder.DefineProperty` can not be located. According to [Unity's compatibility reference](https://docs.unity3d.com/410/Documentation/ScriptReference/MonoCompatibility.html) `DefineProperty` should be there.
 
I'd appreciate any advice on ways I can try to fix this, or on whether this is more likely to be a Unity-specific or a Mono-specific thing.

**Comment by jonorossi** [comment]
@dtchepak sounds like a Unity defect, I've never seen anything so fundamental broken. Our Mono build is successful on every build, I'd try to replicate it using `TypeBuilder` without DynamicProxy.

I'll close this issue for now, but if you do believe it is a DP defect feel free to reopen it.

**Comment by dtchepak** [comment]
@jonorossi Perfect, that's just the info I was hoping for. Thanks Jon!

**Comment by benji7425** [comment]
Considering this is the only result when searching for this error, I hope you don't mind if I hijack it to provide some more information to anyone in future who searches for it.

To those tearing their hair out over this, all is not lost!

I assume you are trying to use NSubstitute in a version of Unity prior to 2018, and have discovered that version 2.0.3 is the highest version with a .NET 3.5 build available on NuGet.
There is one very simple fix: **use version 2.0.1-rc, or 1.10.0.**

If you are using Unity 2018+ (2018.1 currently in beta when writing this), you *may* have some luck using a .NET 4.x build of NSubstitute version 3.x+. You will need to switch your Unity scripting runtime to "Stable (.NET 4.x Equivalent)" via Edit > Project Settings > Player.
I've tried this myself and unfortunately Unity complained that it was a broken assembly, but then this is a beta version, and .NET 4.x compatibility is brand new, so perhaps it will work in future.
